File: src/paper_search_mcp/search/webofscience.py
"""Web of Science searcher — Clarivate WoS Starter API."""

from __future__ import annotations

from datetime import datetime
from typing import Any, List, Optional

from loguru import logger

from ..models import Paper
from .base import BaseSearcher

# WoS field tags for query building
_WOS_FIELD_TAGS = frozenset({
    "TS=", "TI=", "AU=", "SO=", "PY=", "DO=", "IS=", "VL=",
    "PG=", "CS=", "DT=", "PMID=", "FPY=", "DOP=", "AI=",
    "UT=", "OG=", "SUR=",
})


class WebOfScienceSearcher(BaseSearcher):
    """Searcher for Web of Science via Clarivate Starter API.

    Supports v1/v2 API with automatic fallback on server errors.
    Requires an API key (X-ApiKey header).
    """

    platform_name = "webofscience"
    base_url = "https://api.clarivate.com/apis/wos-starter"

    def __init__(self, config: Any) -> None:
        super().__init__(config)
        self._preferred_version: str = "v2"
        self._api_version: str = self._preferred_version
        self._fallback_attempted: bool = False

    @property
    def _api_url(self) -> str:
        return f"{self.base_url}/{self._api_version}"

    def _switch_to_fallback(self) -> bool:
        """Switch API version on failure. Returns False if already tried."""
        if self._fallback_attempted:
            return False
        self._api_version = "v1" if self._api_version == "v2" else "v2"
        self._fallback_attempted = True
        logger.warning(f"[webofscience] Switching to API {self._api_version}")
        return True

    def _reset_fallback(self) -> None:
        """Reset to preferred version after a successful request."""
        if self._fallback_attempted and self._api_version != self._preferred_version:
            self._fallback_attempted = False
            self._api_version = self._preferred_version

    async def search(
        self,
        query: str,
        max_results: Optional[int] = None,
        year_from: Optional[int] = None,
        year_to: Optional[int] = None,
        author: Optional[str] = None,
        journal: Optional[str] = None,
        **kwargs: Any,
    ) -> List[Paper]:
        self.reset_diagnostics(
            query=query,
            request_url=f"{self._api_url}/documents",
        )
        if not self.api_key:
            message = "Web of Science request failed: WOS_API_KEY is not configured."
            self.update_diagnostics(error=message)
            logger.error(f"[webofscience] {message}")
            return []

        max_results = max_results or self.max_results
        wos_query = self._build_query(
            query,
            year_from=year_from,
            year_to=year_to,
            author=author,
            journal=journal,
            doi=kwargs.get("doi"),
            issn=kwargs.get("issn"),
            document_type=kwargs.get("document_type"),
        )
        self.update_diagnostics(query=wos_query)

        params: dict[str, Any] = {
            "q": wos_query,
            "db": kwargs.get("db", "WOS"),
            "limit": max(1, min(max_results, 50)),
            "page": max(1, int(kwargs.get("page", 1))),
            "sortField": self._map_sort_field(kwargs.get("sort_by", "relevance")),
        }

        if kwargs.get("edition"):
            params["edition"] = kwargs["edition"]
        if kwargs.get("detail"):
            params["detail"] = kwargs["detail"]
        if kwargs.get("publish_time_span"):
            params["publishTimeSpan"] = kwargs["publish_time_span"]
        if kwargs.get("modified_time_span"):
            params["modifiedTimeSpan"] = kwargs["modified_time_span"]
        if kwargs.get("tc_modified_time_span"):
            params["tcModifiedTimeSpan"] = kwargs["tc_modified_time_span"]

        headers = {
            "X-ApiKey": self.api_key,
            "Accept": "application/json",
        }

        try:
            data = await self._request_with_fallback(
                f"{self._api_url}/documents", params=params, headers=headers,
            )
            self._reset_fallback()
            papers = self._parse_response(data)
            self.update_diagnostics(
                request_url=f"{self._api_url}/documents",
                status_code=200,
                result_count=len(papers),
                error=None,
            )
            return papers

        except Exception as e:
            status_code = getattr(getattr(e, "response", None), "status_code", None)
            message = self._format_search_error(status_code, e)
            self.update_diagnostics(
                request_url=f"{self._api_url}/documents",
                status_code=status_code,
                error=message,
                exception_type=type(e).__name__,
            )
            logger.error(f"[webofscience] {message}")
            return []

    async def _request_with_fallback(self, url: str, **kwargs: Any) -> Any:
        """Make request with automatic v1/v2 fallback on server errors."""
        try:
            return await self._get_json(url, **kwargs)
        except Exception as e:
            status = getattr(getattr(e, "response", None), "status_code", None)
            should_fallback = (
                not status  # network error
                or status == 404
                or status >= 500
            )
            if should_fallback and self._switch_to_fallback():
                # Retry with new version URL
                new_url = f"{self._api_url}/documents"
                self.update_diagnostics(request_url=new_url)
                return await self._get_json(new_url, **kwargs)
            raise

    @staticmethod
    def _format_search_error(status_code: Optional[int], exc: Exception) -> str:
        if status_code == 401:
            return (
                "Web of Science request failed: 401 Unauthorized. "
                "Check WOS_API_KEY and WoS Starter API entitlement."
            )
        if status_code == 403:
            return (
                "Web of Science request failed: 403 Forbidden. "
                "Check WoS Starter API entitlement and account permissions."
            )
        if status_code == 404:
            return "Web of Science request failed: 404 Not Found."
        if status_code is not None:
            return f"Web of Science request failed: HTTP {status_code}."
        return f"Web of Science request failed: {exc}"

    def _build_query(
        self,
        query: str,
        year_from: Optional[int] = None,
        year_to: Optional[int] = None,
        author: Optional[str] = None,
        journal: Optional[str] = None,
        doi: Optional[str] = None,
        issn: Optional[str] = None,
        document_type: Optional[str] = None,
    ) -> str:
        """Build a WoS field-tag query string."""
        parts: list[str] = []

        # Main topic query
        if query and query.strip():
            has_tag = any(
                query.upper().startswith(tag) or f" {tag}" in query.upper()
                for tag in _WOS_FIELD_TAGS
            )
            if has_tag:
                parts.append(query)
            else:
                parts.append(f"TS=({query})")

        # Year range
        if year_from or year_to:
            start = year_from or 1900
            end = year_to or 2099
            parts.append(f"PY=({start}-{end})")

        # Author
        if author:
            parts.append(f"AU=({author})")

        # Journal / Source
        if journal:
            parts.append(f"SO=({journal})")

        if doi:
            parts.append(f'DO="{doi}"')

        if issn:
            parts.append(f"IS={issn}")

        if document_type:
            doc_types = [item.strip() for item in document_type.split(",") if item.strip()]
            if doc_types:
                joined = " OR ".join(f'"{item}"' for item in doc_types)
                parts.append(f"DT=({joined})")

        return " AND ".join(parts)

    def _map_sort_field(self, sort_by: str) -> str:
        mapping = {
            "relevance": "RS+D",
            "date": "PY+D",
            "citations": "TC+D",
        }
        return mapping.get(sort_by, "RS+D")

    def _parse_response(self, data: dict) -> List[Paper]:
        hits = data.get("hits", [])
        if not isinstance(hits, list):
            return []

        total = data.get("metadata", {}).get("total", 0)
        logger.debug(f"[webofscience] {len(hits)} hits / {total} total")

        papers: List[Paper] = []
        for record in hits:
            paper = self._parse_record(record)
            if paper:
                papers.append(paper)
        return papers

    @staticmethod
    def _format_pages(pages: Any) -> Optional[str]:
        """Normalize WoS page metadata for user-facing output."""
        if isinstance(pages, str):
            return pages or None
        if isinstance(pages, dict):
            if pages.get("range"):
                return str(pages["range"])
            begin = pages.get("begin")
            end = pages.get("end")
            if begin and end:
                return f"{begin}-{end}"
            if pages.get("count"):
                return str(pages["count"])
        return None

    def _parse_record(self, record: dict) -> Optional[Paper]:
        try:
            uid = record.get("uid", "")
            title = record.get("title", "") or "No title"

            # Authors
            authors: list[str] = []
            names = record.get("names", {})
            if isinstance(names, dict):
                author_list = names.get("authors", [])
                authors = [a.get("displayName", "") for a in author_list if a.get("displayName")]

            # Abstract
            abstract = record.get("abstract", "") or ""

            # Source / publication info
            source = record.get("source", {}) or {}
            journal = source.get("sourceTitle", "") or None
            year = source.get("publishYear")
            volume = source.get("volume") or None
            issue = source.get("issue") or None
            pages = source.get("pages") or None

            # DOI
            identifiers = record.get("identifiers", {}) or {}
            doi = identifiers.get("doi") or None

            # Citations
            citation_count = 0
            citations = record.get("citations", [])
            if citations and isinstance(citations, list):
                first = citations[0] if citations else {}
                citation_count = (
                    first.get("citingArticlesCount")
                    or first.get("count")
                    or 0
                )

            # Keywords
            keywords_data = record.get("keywords", {}) or {}
            keywords = keywords_data.get("authorKeywords", []) or []

            # Document types
            types = record.get("types", []) or []

            # URL
            wos_url = f"https://www.webofscience.com/wos/woscc/full-record/{uid}" if uid else ""

            published_date = None
            if isinstance(year, int):
                published_date = datetime(year, 1, 1)

            return Paper(
                paper_id=uid,
                title=title.strip(),
                authors=authors,
                abstract=abstract.strip(),
                doi=doi,
                published_date=published_date,
                url=wos_url,
                source="webofscience",
                categories=types,
                keywords=keywords,
                citations=int(citation_count),
                journal=journal,
                year=year,
                extra={
                    "volume": volume,
                    "issue": issue,
                    "pages": self._format_pages(pages),
                },
            )
        except Exception as e:
            logger.warning(f"[webofscience] Parse error: {e}")
            return None

File: src/providers/web_of_science.rs
use super::{Provider, ProviderError, ProviderSearchResult};
use crate::config::Config;
use crate::error::AppError;
use crate::infra::http::{HttpClient, HttpError, HttpResponse};
use crate::model::{Paper, PaperSearchInput, ProviderName, SortBy, WosSearchOptions};
use async_trait::async_trait;
use chrono::{TimeZone, Utc};
use reqwest::header::{ACCEPT, HeaderMap, HeaderValue};
use serde_json::Value;
use std::collections::BTreeMap;

const BASE_URL: &str = "https://api.clarivate.com/apis/wos-starter";
const FIELD_TAGS: &[&str] = &[
    "TS=", "TI=", "AU=", "SO=", "PY=", "DO=", "IS=", "VL=", "PG=", "CS=", "DT=", "PMID=", "FPY=",
    "DOP=", "AI=", "UT=", "OG=", "SUR=",
];

#[derive(Debug)]
pub struct WebOfScienceProvider {
    http: HttpClient,
    api_key: Option<String>,
    max_results: u32,
}

impl WebOfScienceProvider {
    pub fn new(config: &Config) -> Result<Self, ProviderError> {
        let provider_config = config.provider(ProviderName::Webofscience);
        Ok(Self {
            http: HttpClient::new(ProviderName::Webofscience, provider_config, config)?,
            api_key: provider_config.api_key.clone(),
            max_results: provider_config.max_results,
        })
    }

    async fn request_with_fallback(
        &self,
        params: &[(String, String)],
        headers: HeaderMap,
    ) -> Result<HttpResponse, HttpError> {
        let primary = format!("{BASE_URL}/v2/documents");
        match self.http.get(&primary, params, headers.clone()).await {
            Ok(response) => Ok(response),
            Err(error)
                if error.status_code.is_none()
                    || error.status_code == Some(404)
                    || error.status_code.is_some_and(|status| status >= 500) =>
            {
                let fallback = format!("{BASE_URL}/v1/documents");
                self.http.get(&fallback, params, headers).await
            }
            Err(error) => Err(error),
        }
    }
}

#[async_trait]
impl Provider for WebOfScienceProvider {
    fn name(&self) -> ProviderName {
        ProviderName::Webofscience
    }

    async fn search(
        &self,
        input: &PaperSearchInput,
    ) -> Result<ProviderSearchResult, ProviderError> {
        let api_key = self.api_key.as_ref().ok_or_else(|| ProviderError {
            code: "missing_credentials".into(),
            message: "WOS_API_KEY is required when Web of Science is enabled".into(),
            status_code: None,
            request_url: None,
        })?;
        let query = build_query(input);
        let options = input.wos_options.as_ref().cloned().unwrap_or_default();
        let params = vec![
            ("q".into(), query),
            ("db".into(), "WOS".into()),
            (
                "limit".into(),
                input.max_results.min(self.max_results).min(50).to_string(),
            ),
            ("page".into(), options.page.unwrap_or(1).to_string()),
            ("sortField".into(), sort_field(input.sort_by).into()),
        ];
        let mut headers = HeaderMap::new();
        headers.insert(
            "x-apikey",
            HeaderValue::from_str(api_key).map_err(|_| ProviderError {
                code: "invalid_api_key".into(),
                message: "WOS_API_KEY contains invalid header characters".into(),
                status_code: None,
                request_url: None,
            })?,
        );
        headers.insert(ACCEPT, HeaderValue::from_static("application/json"));
        let response = self.request_with_fallback(&params, headers).await.map_err(|error| {
            let mut provider_error = ProviderError::from(error);
            provider_error.message = match provider_error.status_code {
                Some(401) => "Web of Science returned 401 Unauthorized; check WOS_API_KEY and Starter API entitlement".into(),
                Some(403) => "Web of Science returned 403 Forbidden; check Starter API account permissions".into(),
                Some(404) => "Web of Science Starter API endpoint was not found".into(),
                _ => provider_error.message,
            };
            provider_error
        })?;
        let payload: Value = serde_json::from_str(&response.body).map_err(|error| {
            AppError::Parse(format!("[webofscience] invalid JSON response: {error}"))
        })?;
        let papers = payload
            .get("hits")
            .and_then(Value::as_array)
            .into_iter()
            .flatten()
            .filter_map(parse_record)
            .collect();

        Ok(ProviderSearchResult {
            papers,
            status_code: Some(response.status_code),
            request_url: Some(response.request_url),
            cached: response.cached,
            message: None,
        })
    }
}

fn build_query(input: &PaperSearchInput) -> String {
    let upper = input.query.to_ascii_uppercase();
    let mut parts = vec![if FIELD_TAGS
        .iter()
        .any(|tag| upper.starts_with(tag) || upper.contains(&format!(" {tag}")))
    {
        input.query.clone()
    } else {
        format!("TS=({})", input.query)
    }];
    if input.year_from.is_some() || input.year_to.is_some() {
        parts.push(format!(
            "PY=({}-{})",
            input.year_from.unwrap_or(1900),
            input.year_to.unwrap_or(2099)
        ));
    }
    if let Some(author) = input
        .author
        .as_deref()
        .filter(|value| !value.trim().is_empty())
    {
        parts.push(format!("AU=({author})"));
    }
    if let Some(journal) = input
        .journal
        .as_deref()
        .filter(|value| !value.trim().is_empty())
    {
        parts.push(format!("SO=({journal})"));
    }
    if let Some(options) = &input.wos_options {
        append_options(&mut parts, options);
    }
    parts.join(" AND ")
}

fn append_options(parts: &mut Vec<String>, options: &WosSearchOptions) {
    if let Some(doi) = options
        .doi
        .as_deref()
        .filter(|value| !value.trim().is_empty())
    {
        parts.push(format!("DO=\"{doi}\""));
    }
    if let Some(issn) = options
        .issn
        .as_deref()
        .filter(|value| !value.trim().is_empty())
    {
        parts.push(format!("IS={issn}"));
    }
    if let Some(document_types) = options.document_type.as_deref() {
        let values = document_types
            .split(',')
            .map(str::trim)
            .filter(|value| !value.is_empty())
            .map(|value| format!("\"{value}\""))
            .collect::<Vec<_>>();
        if !values.is_empty() {
            parts.push(format!("DT=({})", values.join(" OR ")));
        }
    }
}

const fn sort_field(sort_by: SortBy) -> &'static str {
    match sort_by {
        SortBy::Relevance => "RS+D",
        SortBy::Date => "PY+D",
        SortBy::Citations => "TC+D",
    }
}

fn parse_record(record: &Value) -> Option<Paper> {
    let uid = string_at(record, &["uid"]);
    let title = string_at(record, &["title"]);
    if uid.is_none() && title.is_none() {
        return None;
    }
    let source = record.get("source").unwrap_or(&Value::Null);
    let identifiers = record.get("identifiers").unwrap_or(&Value::Null);
    let names = record
        .pointer("/names/authors")
        .and_then(Value::as_array)
        .map(|authors| {
            authors
                .iter()
                .filter_map(|author| string_at(author, &["displayName"]))
                .collect()
        })
        .unwrap_or_default();
    let citations = record
        .get("citations")
        .and_then(Value::as_array)
        .and_then(|items| items.first())
        .and_then(|item| {
            item.get("citingArticlesCount")
                .or_else(|| item.get("count"))
        })
        .and_then(Value::as_u64)
        .unwrap_or(0);
    let keywords = record
        .pointer("/keywords/authorKeywords")
        .and_then(Value::as_array)
        .map(|values| {
            values
                .iter()
                .filter_map(Value::as_str)
                .map(str::to_owned)
                .collect()
        })
        .unwrap_or_default();
    let categories = record
        .get("types")
        .and_then(Value::as_array)
        .map(|values| {
            values
                .iter()
                .filter_map(Value::as_str)
                .map(str::to_owned)
                .collect()
        })
        .unwrap_or_default();
    let year = source
        .get("publishYear")
        .and_then(Value::as_i64)
        .and_then(|value| i32::try_from(value).ok());
    let published_date = year.and_then(|year| Utc.with_ymd_and_hms(year, 1, 1, 0, 0, 0).single());
    let uid = uid.unwrap_or_default();
    let mut extra = BTreeMap::new();
    if let Some(value) = source.get("pages").and_then(format_pages) {
        extra.insert("pages_raw".into(), Value::String(value.clone()));
    }

    Some(Paper {
        paper_id: uid.clone(),
        title: title.unwrap_or_else(|| "No title".into()).trim().to_owned(),
        authors: names,
        abstract_text: string_at(record, &["abstract"]).unwrap_or_default(),
        doi: string_at(identifiers, &["doi"]),
        issn: string_at(identifiers, &["issn"]),
        published_date,
        year,
        url: if uid.is_empty() {
            String::new()
        } else {
            format!("https://www.webofscience.com/wos/woscc/full-record/{uid}")
        },
        source: ProviderName::Webofscience,
        sources: vec![ProviderName::Webofscience],
        journal: string_at(source, &["sourceTitle"]),
        categories,
        keywords,
        citations,
        volume: string_at(source, &["volume"]),
        issue: string_at(source, &["issue"]),
        pages: source.get("pages").and_then(format_pages),
        extra,
        ..Paper::default()
    })
}

fn string_at(value: &Value, path: &[&str]) -> Option<String> {
    path.iter()
        .try_fold(value, |current, key| current.get(*key))
        .and_then(Value::as_str)
        .filter(|value| !value.is_empty())
        .map(str::to_owned)
}

fn format_pages(value: &Value) -> Option<String> {
    if let Some(value) = value.as_str() {
        return (!value.is_empty()).then(|| value.to_owned());
    }
    let object = value.as_object()?;
    if let Some(range) = object.get("range").and_then(Value::as_str) {
        return Some(range.to_owned());
    }
    let begin = object.get("begin").and_then(Value::as_str);
    let end = object.get("end").and_then(Value::as_str);
    match (begin, end) {
        (Some(begin), Some(end)) => Some(format!("{begin}-{end}")),
        _ => object.get("count").map(ToString::to_string),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn base_input() -> PaperSearchInput {
        serde_json::from_value(serde_json::json!({"query": "machine learning"})).unwrap()
    }

    #[test]
    fn builds_tagged_query() {
        let mut input = base_input();
        input.year_from = Some(2020);
        input.year_to = Some(2025);
        input.wos_options = Some(WosSearchOptions {
            document_type: Some("Article, Review".into()),
            ..WosSearchOptions::default()
        });
        assert_eq!(
            build_query(&input),
            "TS=(machine learning) AND PY=(2020-2025) AND DT=(\"Article\" OR \"Review\")"
        );
    }

    #[test]
    fn parses_wos_record() {
        let value = serde_json::json!({
            "uid": "WOS:1", "title": "Paper", "abstract": "Text",
            "names": {"authors": [{"displayName": "Alice"}]},
            "source": {"sourceTitle": "Journal", "publishYear": 2025, "pages": {"range": "1-5"}},
            "identifiers": {"doi": "10.1/test", "issn": "1234-5678"},
            "citations": [{"citingArticlesCount": 3}], "types": ["Article"]
        });
        let paper = parse_record(&value).unwrap();
        assert_eq!(paper.citations, 3);
        assert_eq!(paper.journal.as_deref(), Some("Journal"));
        assert_eq!(paper.pages.as_deref(), Some("1-5"));
    }
}
